csrc/flash_attn/fmha_api.cpp

```diff
@@ -385,6 +450,41 @@ mha_bwd(const at::Tensor &dout,  // total_q x num_heads, x head_size
     CHECK_SHAPE(cu_seqlens_q, batch_size + 1);
     CHECK_SHAPE(cu_seqlens_k, batch_size + 1);
 
+    int bias_mod_size = 0;
+    if (attn_bias.has_value()) {
+        TORCH_CHECK(attn_bias.value().is_cuda());
+        TORCH_CHECK(attn_bias.value().dtype() == q_dtype);
+        TORCH_CHECK(attn_bias.value().is_contiguous());
+        // check attn_bias shape
+        const auto bias_sizes = attn_bias->sizes();
+        // last two dimension
+        bias_mod_size = bias_sizes[0];
+        TORCH_CHECK(bias_sizes[1] == num_heads);
+    }
+
+    int mask_head_mod_size = 0;
+    int mask_seq_mod_size = 0;
+    if (attn_mask.has_value()) {
+        TORCH_CHECK(attn_mask.value().is_cuda());
+        TORCH_CHECK(attn_mask.value().dtype() == q_dtype);
+        TORCH_CHECK(attn_mask.value().is_contiguous());
+
+        const auto mask_sizes = attn_mask->sizes();
+        // last two dimension
+        mask_head_mod_size = mask_sizes[1];
+        mask_seq_mod_size = mask_sizes[2];
+        TORCH_CHECK(mask_sizes[1] == 1 || mask_sizes[1] == num_heads);
+        TORCH_CHECK(mask_sizes[2] == 1 || mask_sizes[2] == max_seqlen_q_);
+    }
+
+    auto opts = q.options();
+    at::Tensor ds;
+    if (attn_bias.has_value()) {
+        ds = torch::empty({batch_size, num_heads, max_seqlen_q_, max_seqlen_k_}, opts.dtype(q_dtype));
+        ds.zero_();
```

Comment:
ds?

Reply:
we didn't use the block reduce method to get the `dbias`, so we store the ds and start a new small kernel to compute `dbias`.